### ISSUE TITLE
Splitting up large data sets

### ISSUE THREAD

**Issue by mkovalski** [uber/horovod#166]
Hello,

First off, thanks for creating such a great library, it is really appreciated!

What I wanted to know is if horovod supports a effective way for splitting up datasets between MPI processes or would this need to be prepossessed and sent via some MPI send/recv commands? In the tensorflow_mnist example, it looks as though each process will read the entire dataset. For large datasets, this can be problematic due to memory constraints and threading/queues cannot be utilized. Maybe there's some documentation I missed somewhere.

**Comment by alsrgv** [comment]
@mkovalski, welcome, we're happy it's serving the community!

As for data partitioning, what we typically do is just random sampling from the whole dataset, if it's available via some distributed filesystem (NFS, HDFS, etc).  We then but the number of iterations by the number of workers.

Example of this approach is demonstrated [here](https://github.com/uber/horovod/blob/master/examples/keras_imagenet_resnet50.py).

We compared this approach with pre-partitioning that would ensure that each example is read exactly once per epoch but didn't observe any difference in convergence or final accuracy, and so we stuck with the random sampling approach since it's easier to code in Keras / TF.

**Comment by mkovalski** [comment]
Thanks for the response! Ended up going with this technique and it works well.

**Comment by EhsanM4t1qbit** [comment]
Hi, I was wondering if you could clarify a couple of pints for me please. They are confusing me.

1) In the mnist TF example, each process (and hence each GPU) seems to be working on a separate part of the dataset. 
`mnist = learn.datasets.mnist.read_data_sets('MNIST-data-%d' % hvd.rank())`
Is that right? And then the gradient are averaged at the end of each batch during training, right?

2) In the Keras example you posted above, the data is not explicitly distributed. Who is taking care of that?

** For my application, I have my data saved up as multiple TFRecord files and I want to code in tesnorflow. I am not sure how to distribute the data though. The data is large (40-50 GB). Ideally I'd like to distribute the data batch by batch between all the GPUs, but I don't know how to do that.

Your help would be greatly appreciated.
